modules/widespaceBidAdapter.js

```diff
@@ -0,0 +1,239 @@
+      if (PERF_DATA[i]) {
+        Object.keys(PERF_DATA[i]).forEach((perfDataKey) => {
+          data[perfDataKey] = PERF_DATA[i][perfDataKey];
+        });
+      }
+
+      // Include connection info if available
+      const CONNECTION = navigator.connection || navigator.webkitConnection;
+      if (CONNECTION && CONNECTION.type && CONNECTION.downlinkMax) {
+        data['netinfo.type'] = CONNECTION.type;
+        data['netinfo.downlinkMax'] = CONNECTION.downlinkMax;
+      }
+
+      // Include debug data when available
+      if (!isInHostileIframe) {
+        const DEBUG_AD = (window.top.location.hash.split('&').find((val) => {
+          return val.includes('WS_DEBUG_FORCEADID');
+        }) || '').split('=')[1];
+        data.forceAdId = DEBUG_AD;
+      }
+
+      // Remove empty params
+      Object.keys(data).forEach((key) => {
+        if (data[key] === '' || data[key] === undefined) {
+          delete data[key];
+        }
+      });
+
+      serverRequests.push({
+        method: 'POST',
+        options: {
+          contentType: 'application/x-www-form-urlencoded'
+        },
+        url: REQUEST_SERVER_URL,
+        data: parseQueryStringParameters(data)
+      });
+    });
+    preReqTime = Date.now();
+    return serverRequests;
+  },
+
+  interpretResponse: function(serverResponse, request) {
+    const responseTime = Date.now() - preReqTime;
+    const successBids = serverResponse.body || [];
+    let bidResponses = [];
+    successBids.forEach((bid) => {
+      storeData({
+        'perf_status': 'OK',
+        'perf_reqid': bid.reqId,
+        'perf_ms': responseTime
+      }, `${LS_KEYS.PERF_DATA}${bid.reqId}`);
+      if (bid.status === 'ad') {
+        bidResponses.push({
+          requestId: bid.bidId,
+          cpm: bid.cpm,
+          width: bid.width,
+          height: bid.height,
+          creativeId: bid.adId,
+          currency: bid.currency,
+          netRevenue: Boolean(bid.netRev),
+          ttl: bid.ttl,
+          referrer: getTopWindowReferrer(),
+          ad: bid.code
+        });
+      }
+    });
+
+    return bidResponses
+  },
+
+  getUserSyncs: function(syncOptions, serverResponses = []) {
+    let userSyncs = [];
+    userSyncs = serverResponses.reduce((allSyncPixels, response) => {
+      if (response && response.body && response.body[0]) {
+        (response.body[0].syncPixels || []).forEach((url) => {
+          allSyncPixels.push({type: 'image', url});
+        });
+      }
+      return allSyncPixels;
+    }, []);
+    return userSyncs;
+  }
+};
+
+function storeData(data, name, stringify = true) {
+  const value = stringify ? JSON.stringify(data) : data;
+  if (LOCAL_STORAGE_AVAILABLE) {
+    localStorage.setItem(name, value);
+    return true;
+  } else if (COOKIE_ENABLED) {
+    const theDate = new Date();
+    const expDate = new Date(theDate.setMonth(theDate.getMonth() + 12)).toGMTString();
+    window.document.cookie = `${name}=${value};path=/;expires=${expDate}`;
```

Comment:
This creates a cookie in the publisher's cookie space which seems like something the publisher should at least informed is happening.  I don't see any mention in the docs that adapter creates cookies.  However, I see other adapters doing this too and do not know of specific prohibition against it.

Reply:
Yes I havnt found any adapter doc mentioning it, also cookie is just fallback and we have seen that 99.9% chances are for localStorage to be used.